clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -511,17 +511,18 @@ private static int parseAcks(String acksString) {
      * @return The amount of time we waited in ms
      */
     private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
-        // add topic to metadata topic list if it is not there already.
-        if (!this.metadata.containsTopic(topic))
-            this.metadata.add(topic);
-
+        // add topic to metadata topic list if it is not there already and reset expiry
+        this.metadata.add(topic);
         if (metadata.fetch().partitionsForTopic(topic) != null)
             return 0;
 
         long begin = time.milliseconds();
         long remainingWaitMs = maxWaitMs;
         while (metadata.fetch().partitionsForTopic(topic) == null) {
             log.trace("Requesting metadata update for topic {}.", topic);
+            // Add the topic to metadata before requesting update to ensure that the topic is included
+            // in the metadata topic list since unknown topics are removed from the metadata
+            this.metadata.add(topic);
```

Comment:
@rajinisivaram That's a fair point. It seems a little odd to depend on verbose logging to make the problem more visible, but I'm not too sure. The `TimeoutException` that is thrown from that function is already pretty clear about the nature of the problem, but maybe we could add something to the message to make the specific failure clear?

@ijuma What do you think?